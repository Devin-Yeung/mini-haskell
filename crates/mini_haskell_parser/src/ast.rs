use mini_haskell_diagnostic::span::Span;

#[derive(Debug)]
pub struct Expr {
    pub kind: ExprKind,
    pub span: Span,
}

#[derive(Debug)]
pub enum ExprKind {
    Literal(Literal),
    BinaryExpr(BinaryExpr),
    CondExpr(CondExpr),
}

#[derive(Debug)]
pub enum Literal {
    NatureNum(usize),
    Bool(bool),
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum BinaryOp {
    Plus,
    Ampersand,
    Less,
    Equal,
}

#[derive(Debug)]
pub struct BinaryExpr {
    pub lhs: Box<Expr>,
    pub op: BinaryOp,
    pub rhs: Box<Expr>,
}

#[derive(Debug)]
pub struct CondExpr {
    condition: Box<Expr>,
    then_branch: Box<Expr>,
    else_branch: Box<Expr>,
}

impl Expr {
    pub fn new(kind: ExprKind, span: Span) -> Self {
        Self { kind, span }
    }
}
