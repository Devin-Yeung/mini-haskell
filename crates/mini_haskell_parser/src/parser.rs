use crate::ast::{Expr, Literal};
use crate::error::SyntaxError;
use mini_haskell_lexer::lexer::{TokenTy, Tokenizer};
use std::iter::Peekable;

pub struct Parser<'src> {
    tokenizer: Peekable<Tokenizer<'src>>,
}

impl<'src> Parser<'src> {
    pub fn new<S: AsRef<str> + ?Sized>(src: &'src S) -> Self {
        Self {
            tokenizer: Tokenizer::new(src).peekable(),
        }
    }

    fn consume(&mut self, ty: TokenTy) -> Result<TokenTy, SyntaxError> {
        dbg!(self.tokenizer.peek());
        match self.peek_type()? {
            found if found == ty => Ok(self.advance()?),
            found => Err(SyntaxError::UnexpectedToken {
                expected: ty.name(),
                found: found.name(),
            }),
        }
    }

    fn consume_if(&mut self, ty: TokenTy) -> bool {
        if self.peek_type() == Ok(ty) {
            self.advance().unwrap();
            return true;
        }
        false
    }

    fn peek_type(&mut self) -> Result<TokenTy, SyntaxError> {
        dbg!(self.tokenizer.peek());
        match self.tokenizer.peek() {
            None => Ok(TokenTy::EOF),
            Some(Ok(token)) => Ok(token.ty.clone()),
            Some(Err(_)) => self.advance(),
        }
    }

    fn advance(&mut self) -> Result<TokenTy, SyntaxError> {
        self.tokenizer
            .next()
            .map_or(Err(SyntaxError::UnexpectedEOF), |inner| {
                match inner {
                    Ok(tok) => Ok(tok.ty),
                    Err(_) => {
                        // lexing error is ignored and skip to next valid token
                        loop {
                            match self.tokenizer.peek() {
                                None => return Err(SyntaxError::UnexpectedEOF),
                                Some(Ok(tok)) => return Ok(tok.ty.clone()),
                                Some(Err(_)) => {
                                    self.tokenizer.next();
                                }
                            }
                        }
                    }
                }
            })
    }

    /// parse primary expression according to following rules:
    ///
    /// ```text
    /// primary  â†’ NAT | "T" | "F" | IDENTIFIER ;
    /// ```
    fn primary(&mut self) -> Result<Expr, SyntaxError> {
        match self.peek_type()? {
            TokenTy::BoolLit(b) => Ok(Expr::Literal(Literal::Bool(b))),
            TokenTy::NatLit(n) => Ok(Expr::Literal(Literal::NatureNum(n))),
            TokenTy::Identifier(_) => todo!(),
            _ => Err(SyntaxError::Expected("expression")),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use crate::parser::SyntaxError;
    use crate::parser::TokenTy;
    use mini_haskell_lexer::lexer::TokenTy::NatLit;
    use testsuite::unittest;

    unittest!(advance, |src| {
        let mut parser = Parser::new(src);
        let mut result = Vec::<Result<TokenTy, SyntaxError>>::new();
        loop {
            match parser.advance() {
                Err(SyntaxError::UnexpectedEOF) => break,
                Ok(tok) => result.push(Ok(tok)),
                Err(err) => result.push(Err(err)),
            }
        }
        insta::assert_debug_snapshot!(result);
    });

    unittest!(consume, |src| {
        let mut parser = Parser::new(src);
        let result = vec![
            parser.peek_type(),
            parser.consume(NatLit(1)),
            parser.peek_type(),
            parser.consume(NatLit(2)),
            parser.peek_type(),
        ];
        insta::assert_debug_snapshot!(result);
    });

    unittest!(primary, |src| {
        let asts = src
            .split('\n')
            .map(|line| Parser::new(line).primary())
            .collect::<Vec<_>>();
        insta::assert_debug_snapshot!(asts);
    });
}
