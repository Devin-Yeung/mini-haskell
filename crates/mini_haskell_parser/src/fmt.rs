use crate::ast::{BinaryExpr, BinaryOp, CondExpr, Expr, ExprKind, Literal};
use debug_tree::scoped_branch::ScopedBranch;
use debug_tree::TreeBuilder;

pub struct AstViewBuilder {
    builder: TreeBuilder,
    root: ScopedBranch,
}

impl AstViewBuilder {
    pub fn new() -> Self {
        let builder = TreeBuilder::new();
        let root = builder.add_branch("AST");
        AstViewBuilder { builder, root }
    }

    pub fn node(&mut self, name: &str, node: &dyn AstView) -> &mut Self {
        let mut branch = self.builder.add_branch(name);
        node.view(self);
        branch.release();
        self
    }

    pub fn leaf(&mut self, value: &str) -> &mut Self {
        self.builder.add_leaf(value);
        self
    }

    pub fn string(&mut self) -> String {
        self.root.release();
        self.builder.string()
    }
}

pub trait AstView {
    fn view(&self, f: &mut AstViewBuilder);

    fn ast(&self) -> String {
        let mut builder = AstViewBuilder::new();
        self.view(&mut builder);
        builder.string()
    }
}

impl AstView for Expr {
    fn view(&self, f: &mut AstViewBuilder) {
        match &self.kind {
            ExprKind::Literal(lit) => lit.view(f),
            ExprKind::BinaryExpr(expr) => expr.view(f),
            ExprKind::CondExpr(expr) => expr.view(f),
        }
    }
}

impl AstView for Literal {
    fn view(&self, f: &mut AstViewBuilder) {
        match self {
            Literal::NatureNum(n) => {
                f.leaf(&format!("Nat({n})"));
            }
            Literal::Bool(b) => {
                f.leaf(&format!("Bool({b})"));
            }
        }
    }
}

impl AstView for BinaryExpr {
    fn view(&self, f: &mut AstViewBuilder) {
        f.node("lhs", &self.lhs)
            .leaf(&format!("ops: {}", self.op.raw()))
            .node("rhs", &self.rhs);
    }
}

impl AstView for CondExpr {
    fn view(&self, f: &mut AstViewBuilder) {
        f.node("condition", &self.condition)
            .node("then-branch", &self.then_branch)
            .node("else-branch", &self.else_branch);
    }
}

impl AstView for Box<Expr> {
    fn view(&self, f: &mut AstViewBuilder) {
        AstView::view(self.as_ref(), f);
    }
}

impl BinaryOp {
    fn raw(&self) -> &'static str {
        match self {
            BinaryOp::Plus => "+",
            BinaryOp::Ampersand => "&",
            BinaryOp::Less => "<",
            BinaryOp::Equal => "=",
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::fmt::AstView;
    use crate::parser::Parser;
    use testsuite::unittest;

    unittest!(expr, |_, src| {
        let ast = Parser::new(&src).conditional().unwrap().ast();
        insta::assert_snapshot!(ast);
    });
}
