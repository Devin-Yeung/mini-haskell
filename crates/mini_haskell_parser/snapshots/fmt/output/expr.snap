---
source: crates/mini_haskell_parser/src/fmt.rs
expression: ast
---
AST
├╼ condition
│ └╼ Bool(false)
├╼ then-branch
│ └╼ Nat(0)
└╼ else-branch
  ├╼ lhs
  │ ├╼ lhs
  │ │ └╼ Nat(1)
  │ ├╼ ops: <
  │ └╼ rhs
  │   └╼ Nat(0)
  ├╼ ops: &
  └╼ rhs
    ├╼ lhs
    │ └╼ Nat(2)
    ├╼ ops: =
    └╼ rhs
      └╼ Nat(3)
